packages/block-editor/src/components/writing-flow/index.js

```diff
@@ -306,6 +306,16 @@ export default function WritingFlow( { children } ) {
 	}
 
 	function onKeyDown( event ) {
+		// If the writing flow component is inside a parent that is hidden to
+		// assistive technologies, ignore the event. This typically happens
+		// when a modal is opened and `aria-hidden` is applied to an element
+		// wrapping this one. When that happens, keyboard input events (like
+		// those handled here) that change the focus can cause a modal to close
+		// unexpectedly. Returning early avoids that.
+		if ( container.current && container.current.closest( '[aria-hidden]' ) ) {
```

Comment:
aria-hidden can be false right?

Reply:
For me ideally this is not the responsibility of the WritingFlow component because imagine other similar components handling keyboard interactions but more of the component applying aria-hidden. I don't know if that's possiible but we have a "Disabled" component to prevent interactions like this, why not use it instead of just adding aria-hidden.

Reply:
I'd tend to agree. I don't really see the connection between `aria-hidden` and whether events are handled by this component. Part of this is that I don't really understand why there would be typing happening inside an `aria-hidden` element (why is it `aria-hidden`? is it not an accessibility concern that we have interactive content which is hidden from users of assistive technology?). 

Related: https://www.w3.org/TR/using-aria/#fourth

>Applying aria-hidden to a parent/ancestor of a visible interactive element will also result in the interactive element being hidden, so don't do this either:

Reply:
Should there be focus is an element that's hidden? To me it sounds like the problem is that there's still focus in this element, while probably there shouldn't be?

Reply:
> Part of this is that I don't really understand why there would be typing happening inside an aria-hidden element (why is it aria-hidden? is it not an accessibility concern that we have interactive content which is hidden from users of assistive technology?).

The modal adds `aria-hidden` to other content when it's opened using this `ariaHelper.hideApp`.

https://github.com/WordPress/gutenberg/blob/c079120ce39044e53ea41a6a1a246aeec8195433/packages/components/src/modal/index.js#L84-L87

To me, it doesn't seem right that `WritingFlow` is focusing elements that are `aria-hidden`.

> Should there be focus is an element that's hidden? To me it sounds like the problem is that there's still focus in this element, while probably there shouldn't be?

It's not that there's focus on a hidden element. Instead the modal bubbles keyboard events up to WritingFlow. The modal is rendered in a portal, but events still bubble up through portals. Perhaps the solution is to stop that happening, it just seems like we're adding `stopPropagation` pretty liberally at the moment on block UI that uses popovers and modals.

This could be where I'd add `stopPropagation`:
https://github.com/WordPress/gutenberg/blob/master/packages/components/src/modal/frame.js#L69-L73

Reply:
I think I understand now what you're saying and what's being implemented here. I'd also agree with you that I think this is a larger problem of how we deal with events propagation out of popovers (modals, etc). We have a lot of instances of [this sort of code](https://github.com/WordPress/gutenberg/blob/bc5754797cc49ed294ec745bcb19fe053ec34331/packages/block-editor/src/components/link-control/search-input.js#L23-L26) where ideally the code itself shouldn't need to be aware of its ancestors. This example is also interesting in how in trying to deal with one ancestor's handling (`ObserveTyping`), it breaks another (`Popover`'s intention to close on <kbd>Escape</kbd> press).

I mentioned it recently to @youknowriad in a pull request I'm now unable to relocate, but I think it's something where we could even consider being as extreme as to prevent all event propagation out of Popover entirely, treating it as something entirely separate from the rest of the DOM/React hierarchy (the former is already true).

Options here could include:

- Remove [`bubblesVirtually`](https://github.com/WordPress/gutenberg/blob/605ee4df1754c50f47922b2854f364b0f2bcfccc/packages/components/src/popover/index.js#L558) from the rendered Popover slot, which causes it to bubble its events as DOM events and not through React's own virtual event handler. I think this aligns quite well conceptually with what we're trying to do (separate in DOM, detached in React), but I think we've tended to try to move away from and eventually deprecate `bubblesVirtually` so that we have a single uniform rendering behavior of Slot/Fill.
- Mimic the _behavior_ of `bubblesVirtually`, custom to Popover, where it creates its own React hierarchy.
   - In other words, something like `React.render( this.props.children, containerRef )` in Popover itself
   - A big gotcha here usually is making sure some values pass down through to the children, e.g. React context
- Capture and stop propagation of all events
   - It's not clear to me how this could even be done
- Capture and stop propagation of specific events, like `onKeyDown`
   - It could be difficult (but worthwhile) to consider which events this should 
   - Likely the most pragmatic approach

Reply:
I personally think we should still use React event/context bubbling. If you render a Modal in a Block, you expect it to have access to the BlockEditContext.

I feel stopping event propagation for Modals could be considered good since the essence of Modals is that they disable everything on the screen except the Modal itself.

Reply:
Is `Modal` its own thing separate from `Popover`? I assumed this was just a subclass of `Popover`, just like `Dropdown`, etc. It doesn't seem like they ought to be treated any differently.

In any case, I feel like it should be the same in both.

Interesting, in trying to sort through the code of Modal, I note that it's already trying to do something to this effect using `createPortal` as a way to "stop mouse events". I'm not exactly sure how this could possibly work, since my understanding is that `createPortal` would still allow events to bubble to ancestors.

https://github.com/WordPress/gutenberg/blob/0470286ba979ecdb329d890e477e4df3160a521a/packages/components/src/modal/index.js#L126-L128

Reply:
I think Modal has different semantics than Dropdowns and Popovers because it's meant to disable all interactions on the screen when it's opened (that's why it adds aria-hidden everywhere outside the modal), while for the two others, you're still expected to be able to interact with other parts of the UI.

Reply:
> while for the two others, you're still expected to be able to interact with other parts of the UI.

I'm a bit confused on this point. When would we expect interactions elsewhere when a Popover is opened? I mean, I grant that it makes sense that those things are still reachable, but the dialog remains open only so long as focus remains within it. It ceases to remain relevant as soon as interactions (focus, clicks) occur elsewhere. As far as that impacts the implementation, we close it.

I think it's an important point to clarify though, since it's especially relevant for what I'm proposing.

The whole idea with `stopPropagation` at the Popover is under the assumption that, when you're typing within a popover (like in a [link input](https://github.com/WordPress/gutenberg/blob/bc5754797cc49ed294ec745bcb19fe053ec34331/packages/block-editor/src/components/link-control/search-input.js#L23-L26)), you are _explicitly not_ interacting with the rest of the page. And that's the primary reason I think it's acceptable to stop the propagation.

Reply:
Alternatively, we could do something more like what was originally proposed in this pull request, where we reverse how the event is considered to be handled by `ObserveTyping`, `WritingFlow`, etc:

Rather than stop propagation of the event from the popover...
...we consider from those wrappers whether the event occurred _within its own DOM hierarchy_.

This would optimize for an assumption that, if events aren't meant to be handled by these wrappers, then they should occur elsewhere in the DOM.

It's maybe a strange requirement, since the default behavior of Popover will render its contents in-place ([source](https://github.com/WordPress/gutenberg/blob/0470286ba979ecdb329d890e477e4df3160a521a/packages/components/src/popover/index.js#L535-L539)), and this is intended to be a valid usage. But even then, it still seems like we would not want events to be handled outside the context of the Popover. There are interesting parallels in considering that the fact we use Slot/Fill in the first place is largely to avoid conflicts in the real-world implications of in-place rendering (where CSS positioning is difficult taking into account ancestor positioning). Event bubbling behavior is not too much unlike this, so considering "rendering outside the DOM hierarchy" as the solution to "I don't want events to be handled by ancestors" fits under the reasons we use Slot/Fill already.

Reply:
>Rather than stop propagation of the event from the popover...
>...we consider from those wrappers whether the event occurred within its own DOM hierarchy.

See proposed implementation: #19879

Reply:
> I'm a bit confused on this point. When would we expect interactions elsewhere when a Popover is opened? I mean, I grant that it makes sense that those things are still reachable, but the dialog remains open only so long as focus remains within it.

When a popover is opened, you can click on a button outside the popover to trigger it. it's not the case for modals.

Reply:
I think your PR is good regardless because these two components should work on the DOM tree not the React Tree (I attempted the same for WritingFlow at some point but failed).

That said, this doesn't mean the "Modal" shouldn't stop propagation. I don't see these two as mutually exclusive.